src/Engine/Spells/CastSpellInfo.cpp

```diff
@@ -261,49 +290,38 @@ void CastSpellInfoHelpers::castSpell() {
                     pParty->bTurnBasedModeOn) {
                 ++pTurnEngine->pending_actions;
             }
+            // TODO(Nik-RE-dev): recovery time for melee used?
+            setSpellRecovery(pCastSpell, pPlayer->GetAttackRecoveryTime(false));
         } else if (pCastSpell->uSpellID == SPELL_WATER_TOWN_PORTAL) {
-            if (pPlayer->CanCastSpell(uRequiredMana)) {
-                int success_chance_percent = 10 * spell_level;
-                if (spell_mastery != PLAYER_SKILL_MASTERY_GRANDMASTER) {
-                    if (pParty->GetRedOrYellowAlert() ||
-                            grng->Random(100) >= success_chance_percent) {
-                        spellFailed(pCastSpell, LSTR_SPELL_FAILED);
-                        continue;
-                    }
+            int success_chance_percent = 10 * spell_level;
+            bool castSuccessful = true;
+            if (spell_mastery != PLAYER_SKILL_MASTERY_GRANDMASTER) {
+                if (pParty->GetRedOrYellowAlert()) {
+                    spellFailed(pCastSpell, LSTR_SPELL_FAILED);
+                    castSuccessful = false;
+                } else if (grng->Random(100) >= success_chance_percent) {
+                    spellFailed(pCastSpell, LSTR_SPELL_FAILED);
+                    // Mana was not spend before on failure.
+                    pPlayer->SpendMana(uRequiredMana);
+                    castSuccessful = false;
                 }
-                pEventTimer->Pause();
-                TownPortalCasterId = pCastSpell->uPlayerID;
-                pCurrentFrameMessageQueue->AddGUIMessage(UIMSG_OnCastTownPortal, 0, 0);
-                spell_sound_flag = true;
-            } else {
-                pAudioPlayer->PlaySound(SOUND_spellfail0201, 0, 0, -1, 0, 0);
-                pCastSpell->uSpellID = SPELL_NONE;
-                continue;
             }
+            pEventTimer->Pause();
+            TownPortalCasterId = pCastSpell->uPlayerID;
+            pCurrentFrameMessageQueue->AddGUIMessage(UIMSG_OnCastTownPortal, 0, 0);
+            pAudioPlayer->PlaySpellSound(pCastSpell->uSpellID, PID_INVALID);
+            setSpellRecovery(pCastSpell, recoveryTime);
```

Comment:
Maybe this should go to the handler for `UIMSG_ClickTownInTP` event? This is where mana is subtracted.